### DISCUSSION TITLE
엔티티 매니저에 대해 설명해주세요.

### DISCUSSION THREAD

**Discussion by nak-honest** [woowacourse-6th-computer-sign-pen/comsa-interview-study#11] [JPA] [answered]
엔티티 매니저에 대해 설명해주세요.

**Comment by slimsha2dy** [comment, marked as answer]
**답변자: 알파카(최휘용)**

질문자: Entity Manager는 무엇이고, 어떤 역할을 하나요?

답변자: Entity Manager는 엔티티를 관리하는 역할을 합니다.

질문자: 자세히 설명해 주시겠어요?

답변자: 좀 더 정확히 말하면, 엔티티를 영속화하고, 영속화된 엔티티는 영속성 컨텍스트에 들어갑니다. Entity Manager는 이 영속성 컨텍스트 내의 엔티티들을 관리하고, 그 상태를 변경하는 역할을 합니다.

질문자: 상태를 변경한다는 말이 좋네요. 더 설명하실 부분이 있을까요?

답변자: 혹시 구체적으로 궁금하신 부분이 있을까요?

질문자: 그럼 Entity Manager와 영속성 컨텍스트의 차이점은 무엇인가요?

답변자: 둘은 완전히 다른 역할을 합니다. 영속성 컨텍스트는 엔티티들이 매번 DB에 직접 접근하지 않고, 컨텍스트 내에서 관리되도록 합니다. Entity Manager는 이러한 엔티티들을 영속성 컨텍스트로 넣거나, DB에서 가져오거나, 값을 변경하는 역할을 합니다.

질문자: 좋습니다. 그럼 Entity Manager에서 쓰기 지연은 어떻게 동작하나요?

답변자: 쓰기 지연은 영속성 컨텍스트의 특징에 가깝습니다. 내부 구현을 완전히 알지는 못하지만, 업데이트가 발생했을 때 매번 DB에 반영하면 디스크 I/O가 많이 발생합니다. 그래서 변경 사항을 Entity Manager의 1차 캐시와 쓰기 지연 저장소에 저장해 두고, 값을 변경하면 1차 캐시만 수정합니다. 이후 플러시(flush)를 통해 동기화할 때 쓰기 지연 저장소의 모든 변경 사항을 DB에 반영합니다.

질문자: 맞습니다.

답변자: 다른 질문 있으실까요?

질문자: 만약 여러 스레드, 예를 들어 알파카와 라마가 동시에 Entity Manager에 접근한다면 동시성이 보장되나요?

답변자: Entity Manager는 스레드 세이프해야 하지 않을까요?

질문자: 실제로는 보장되지 않습니다. JPA 명세서에 따르면 Entity Manager는 스레드 세이프하지 않다고 합니다.

답변자: 그렇군요. 그렇다면 다른 스레드에서 접근하면 동시성이 보장되지 않는 건가요? 하지만 제가 알기로는 영속성 컨텍스트는 여러 스레드에서 접근할 수 있는 것으로 알고 있는데요.

질문자: 직접 접근하면 동시성이 보장되지 않으니, 스레드 세이프로 만들어서 사용하라는 권장이 있습니다.

답변자: 저희가 보통 사용할 때 직접 접근하지 않는데, 이미 스레드 세이프로 구현되어 있는 건가요? JPA를 직접 사용하지 않고 Hibernate를 사용하니 Hibernate 내부적으로 처리하는 걸까요?

질문자: 스프링에서는 Entity Manager를 프록시로 감싸서, 이 프록시를 통해 Entity Manager를 생성하기 때문에 스레드 세이프를 보장한다고 합니다.

답변자: 프록시를 사용하면 어떻게 스레드 세이프가 되나요?

질문자: 각 프록시 Entity Manager가 서로 다른 스레드에게 다른 영속성 컨텍스트를 제공하여 동시성을 적절히 처리합니다.

답변자: 이해했습니다. 다른 질문 있으실까요?

질문자: 실제 서비스 로직을 작성할 때 Entity Manager를 직접 호출하지는 않죠. 내부적으로는 프록시를 이용해 사용하고 있습니다.

답변자: 내용이 생각보다 간단하군요.

질문자: 그럼 JPA의 EntityManagerFactory와 EntityManager의 차이점은 무엇인가요?

답변자: EntityManagerFactory는 Entity Manager를 생성하는 팩토리 역할을 합니다.

질문자: 맞습니다.

질문자: 영속성 컨텍스트의 장점은 무엇인가요?

답변자: 가장 큰 장점은 1차 캐시와 쓰기 지연입니다.

질문자: 좋습니다.

답변자: 목이 좀 마르네요.

질문자: Entity Manager의 detach 메서드에 대해 설명해 주실 수 있나요?

답변자: detach를 직접 사용해 본 적은 없습니다. 직접 호출하는 경우가 있나요?

질문자: 대용량 엔티티를 처리하는 배치 작업에서 메모리 부족을 방지하기 위해 일정 주기로 detach를 호출할 수 있습니다.

답변자: 그렇군요. 몰랐습니다.

질문자: 쿼리를 실행하면 자동으로 호출된다고도 하네요.

질문자: 그럼 flush와 persist에 대해 설명해 주시겠어요?

답변자: flush는 영속성 컨텍스트와 DB를 동기화하는 역할을 합니다.

질문자: flush를 호출하면 어떤 일이 일어나나요?

답변자: 쓰기 지연 저장소에 있는 변경 사항들이 DB에 반영됩니다.

질문자: 그럼 flush와 commit의 차이는 무엇인가요?

답변자: commit은 트랜잭션이 완전히 끝날 때 수행되며, flush는 변경 사항을 DB에 반영하지만 트랜잭션은 유지됩니다. 따라서 flush 후에도 롤백이 가능합니다.

질문자: 맞습니다. commit은 트랜잭션이 종료될 때 자동으로 이루어지므로, 명시적으로 호출할 일은 거의 없습니다.

답변자: 이해했습니다.

질문자: 혹시 FlushMode에 대해 아시나요?

답변자: FlushMode요?

질문자: FlushMode.AUTO, MANUAL, ALWAYS 등이 있습니다. 트랜잭션을 readOnly로 설정하면 FlushMode가 MANUAL로 바뀝니다. MANUAL에서는 직접 flush를 호출하지 않는 이상 동기화가 이루어지지 않으므로, 읽기 전용에서 사용할 수 있습니다.

답변자: 흥미롭네요.

질문자: 면접에서 활용하실 수 있을 거예요. readOnly가 어떻게 동작하는지, commit이 호출되지 않아서 동기화가 안 되는 건지, flush가 안 되는 건지 등에 대해 설명해 보세요.

질문자: Entity Manager에서 지연 로딩이 발생할 때 성능 이슈가 생길 수 있나요?

답변자: 지연 로딩은 성능상의 문제가 발생할 수 있습니다. 예를 들어, 1대 다 관계 매핑에서 "1" 측을 호출하고 나중에 실제 값을 접근하려고 하면, 그때 쿼리가 실행됩니다. 이때 값을 순회하며 접근하면 N+1 문제가 발생할 수 있습니다.

질문자: 다른 해결 방법은 없을까요?

답변자: 배치 페치를 사용하여 한 번에 데이터를 가져오거나, FetchType.EAGER로 설정하여 지연 로딩을 피할 수 있지만, 이는 다른 문제를 야기할 수 있습니다.

질문자: FetchType.EAGER로 하면 N+1 문제가 발생하지 않을까요?

답변자: 확신할 수는 없습니다.

질문자: 위험할 수 있겠네요.

질문자: EntityManagerFactory와 EntityManager의 차이에 대해 간단히 답변해 주셨는데요. 그럼 Entity Manager는 어떻게 엔티티의 동일성을 보장하나요?

답변자: 엔티티의 동일성은 1차 캐시에서 보장됩니다.

질문자: 어떻게 작동하는지 설명해 주실 수 있나요?

답변자: 정확히는 모르지만, 엔티티의 ID를 기준으로 관리하는 것 같습니다.

질문자: 관련된 내용이 있었던 것 같네요.

질문자: 우리가 엔티티를 사용할 때, 직접 ID를 설정하고 저장하면 어떻게 동작하나요? 예를 들어, 엔티티에 ID를 직접 넣고 persist를 호출하면요.

답변자: 아마도 insert가 될 것 같습니다. 하지만 DB에 해당 ID가 이미 있다면 예외가 발생하거나 업데이트가 될 수 있습니다.

질문자: 저도 정확히는 모르겠습니다.

질문자: 업데이트는 기존 레코드를 수정할 때 사용하죠.

질문자: 만약 ID에 유니크 제약 조건이 없다면, 삽입이 되겠네요.

질문자: 그렇겠죠.
